src/query_plan/fetch_dependency_graph.rs

```diff
@@ -347,18 +686,564 @@ impl DeferTracking {
+    if let QueryGraphEdgeTransition::InterfaceObjectFakeDownCast { .. } = &edge.transition {
+        // We shouldn't add the operation "as is" as it's a down-cast but we're "faking it".
+        // However, if the operation has directives, we should preserve that.
+        let OpPathElement::InlineFragment(inline) = updated_operation else {
+            return Err(FederationError::internal(format!(
+                "Unexpected operation {updated_operation} for edge {edge}"
+            )));
+        };
+        if !inline.data().directives.is_empty() {
+            // We want to keep the directives, but we clear the condition
+            // since it's to a type that doesn't exists in the subgraph we're currently in.
+            updated.node_path = updated
+                .node_path
+                .add(Arc::new(inline.with_updated_type_condition(None).into()))?;
+        }
+    } else {
+        updated.node_path = updated.node_path.add(Arc::new(updated_operation))?;
+    }
+    Ok(updated)
+}
+
+fn wrap_input_selections(
+    _wrapping_type: &CompositeTypeDefinitionPosition,
+    _selections: &NormalizedSelectionSet,
+    _context: &OpGraphPathContext,
+) -> NormalizedSelectionSet {
+    todo!() // Port `wrapInputsSelections` in `buildPlan.ts`
+}
+
+fn compute_input_rewrites_on_key_fetch(
+    _input_type_name: &str,
+    _dest_type: &CompositeTypeDefinitionPosition,
+) -> Option<Vec<Arc<FetchDataRewrite>>> {
+    todo!() // Port `computeInputRewritesOnKeyFetch`
+}
+
+fn create_fetch_initial_path(
+    _supergraph_schema: &ValidFederationSchema,
+    _dest_type: &CompositeTypeDefinitionPosition,
+    _new_context: &OpGraphPathContext,
+) -> Arc<OpPath> {
+    todo!() // Port `createFetchInitialPath`
+}
+
+fn extract_defer_from_operation(
+    _dependency_graph: &mut FetchDependencyGraph,
+    _operation: &OpPathElement,
+    _defer_context: &DeferContext,
+    _node_path: &FetchDependencyGraphNodePath,
+) -> (Option<OpPathElement>, DeferContext) {
+    todo!() // Port `extractDeferFromOperation`
+}
+
+fn handle_requires(
     _dependency_graph: &mut FetchDependencyGraph,
-    _tree: &OpPathTree,
-    _start_node: NodeIndex,
-    _initial_node_path: FetchDependencyGraphNodePath,
-    _initial_defer_context: DeferContext,
-    _initial_conditions: OpGraphPathContext,
-) -> Vec<NodeIndex> {
-    // TODO: port `computeGroupsForTree` in `query-planner-js/src/buildPlan.ts`
-    todo!()
+    _edge_id: EdgeIndex,
+    _requires_conditions: &OpPathTree,
+    (_node_id, _node_path): (NodeIndex, &FetchDependencyGraphNodePath),
+    _context: &OpGraphPathContext,
+    _defer_context: &DeferContext,
+    _created_nodes: &mut IndexSet<NodeIndex>,
+) -> Result<(NodeIndex, FetchDependencyGraphNodePath), FederationError> {
+    // PORT_NOTE: instead of returing IDs of created nodes they should be inserted directly
+    // in the `created_nodes` set passed by mutable reference.
+    todo!() // Port `handleRequires`
 }
```

Comment:
Porting `handle_requires` is tracked at https://github.com/apollographql/federation-next/issues/153